src/components/select/select.js

```diff
@@ -0,0 +1,304 @@
+    error: '',
+    autoFocus: false,
+  };
+
+  state = {
+    textValue: '',
+    selectedOptions: [],
+    isDropDownOpen: this.props.type === TYPE.SEARCH,
+    ...this.props,
+  };
+
+  constructor(props: PropsT) {
+    super(props);
+    if (props.type === TYPE.SELECT) {
+      const {selectedOptions} = this.state;
+      const {options = []} = props;
+      selectedOptions.forEach(selectedOption => {
+        // selected option is not in all options and needs to be added on top
+        if (!options.find(selected => selected.id === selectedOption.id)) {
+          options.unshift({
+            id: selectedOption.id,
+            label: this.getOptionLabel(selectedOption),
+          });
+        }
+      });
+    }
+  }
+
+  componentDidMount() {
+    const document = require('global/document');
+    document.addEventListener('click', this.handleClickEvent, {
+      capture: true,
+    });
+  }
+
+  componentWillUnmount() {
+    const document = require('global/document');
+    document.removeEventListener('click', this.handleClickEvent, {
+      capture: true,
+    });
+  }
+
+  handleClickEvent = (event: MouseEvent) => {
+    // eslint-disable-next-line react/no-find-dom-node
+    const el = findDOMNode(this);
+    // $FlowFixMe
+    if (el && !el.contains(event.target)) {
+      this.setState({isDropDownOpen: false});
+    }
+  };
+
+  onFocus = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onBlur = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onMouseEnter = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onMouseLeave = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onClearAll = (e: SyntheticEvent<HTMLInputElement>) => {
+    const newTextValue = '';
+    this.setState({textValue: newTextValue});
+    this.props.onClearAll(e, {textValue: newTextValue});
+  };
+
+  onSelect = (
+    e: SyntheticEvent<HTMLInputElement>,
+    id: string,
+    label: React$Node,
+  ) => {
+    const {multiple} = this.props;
+    const selected = this.state.selectedOptions.find(tag => tag.id === id);
+    let selectedOptions;
+    if (!selected) {
+      const selectedOptions = multiple
+        ? this.state.selectedOptions.slice()
+        : [];
+      selectedOptions.push({id, label});
+      this.setState({selectedOptions: selectedOptions});
+      if (this.props.type === TYPE.SELECT && !multiple) {
+        this.setState({isDropDownOpen: false});
+      }
+      this.props.onSelect(e, {id, label, selectedOptions});
+    } else if (multiple) {
+      selectedOptions = this.state.selectedOptions.filter(
+        selectedOption => selectedOption.id !== selected.id,
+      );
+      this.setState({
+        selectedOptions: selectedOptions,
+      });
+      this.props.onUnSelect(e, {id, label, selectedOptions});
+    }
+  };
+
+  onKeyUp = (e: SyntheticInputEvent<HTMLInputElement>) => {
+    const newTextValue = e.target.value;
+    this.setState({textValue: newTextValue});
+    this.props.onKeyUp(e, {textValue: newTextValue});
+  };
```

Comment:
I think the Select element should do the option filtering, similar to react-select. IMO that's one of the biggest benefits of a reusable autocomplete component – customers ideally shouldn't have to implement that filtering logic on their own.

Reply:
But all options are provided from outside as it's totally custom thing, there might be ajax call or other logic to recalculate all options. So at the moment of `keyUp` we just don't have actual and correct options to filter. Moreover, whatever is in text value in most commercial implementation (for instance lookup for product) is not exact text to look for, it's just a key and exact algorithm to find and filter results is known only to special search engine logic

Reply:
The vast majority of the time I use select, I have a list of many options, and I just want basic lowercase string matching capability (maybe tokenized by word). This is exactly with react-select offers out of the box, we shouldn't burden customers by making them implement this logic themselves every time they use a select component.

Also, I posted this in a separate thread, but I think for async options we chould accept `options` with the following signature: `(query: string) => Promise<Option[]>`. In this case it would skip our naive string filtering and call the users `options` callback with the current query. We show a loading indicator until the promise resolves.

Reply:
you try to put too many concerns in one place. From one side you want a simple filter, but as I mentioned we have a mixed select\search control, and if it works for select mode, it won't work for search mode. If customers want primitive lowercase search, it is a one liner function, I don't think we need to implement this and support conditionally. On the other hand supporting promise signature also seems to much for me, as remember, it is just a control, ot shouldn't comprehend all logic delivering data to it.

Reply:
> you try to put too many concerns in one place.

This is a fair point. I don't mean to advocate for shoving all this logic into `<Select>`, I mostly just want a solution that works out of the box. That could mean that in the future we split some of this behavior into a separate wrapper component like `<AsyncSelect>`, [which is what react-select does](https://react-select.com/async).

> If customers want primitive lowercase search, it is a one liner function

This kind of oversimplifies the issue a bit in my opinion. Here's the minimum amount of code I would need to write to support this on my own:

```js
class UserSelect extends React.Component {
  state = {
    filteredOptions: this.props.options,
  };
  onKeyUp(params) {
    this.setState({
      filteredOptions: this.props.options.filter(o => (
        o.label.toLowerCase().indexOf(params.textValue.toLowerCase()) >= 0;
      ));
    });
  }
  clearFilter() {
    this.setState({
      filteredOptions: this.props.options
    });
  }
  render() {
    return <Select 
      options={this.filteredOptions}
      onKeyUp={this.onKeyUp} 
      onClear={this.clearFilter}
      onSelect={this.clearFilter} />;
  }
}
```

While this isn't a gigantic amount of code, it does require at least one extra piece of state, I have to subscribe to all three onKeyUp/onClear/onSelect, and in practice should would probably need to be debounced or have better keyword matching. It would get old re-implementing this every time I use a select in my app. I believe this is why react-select offers this behavior out of the box.

If the concern is giving too many concerns to one component, then maybe we keep Select dumb, but also offer an AutocompleteSelect export that has this behavior built in.

It might be interesting to get feedback from some other potential customers too. I think I'm the only one that's reviewed this RFC/PR so far.

Reply:
ok, here are some solutions for your code
1) To avoid extra state we can still use `options`, but they will be not so efficient as options are wiped every time we clean up. To address that we can have some inner cache strategy. However, `filtered` is not needed here
2) To not subscribe to so many we may do on `onChange` handler at least for `clearAll`\`onSelect`\`onUnselect` and provide extra action telling what happened instead like select, unselect, clear (btw, we don't have `clearfilter` right now for this control as Lily wants this X button to clean selected options only)